File: Python/findPoles.py
import numpy as np
import scipy.signal as sig


def findPoles(x, y, initPoles, t, tol=1e-3, iterator=0):
    """Function that performs vector fitting

    Input:
        x is the input data to the system
        y is the response to the input x
        initPoles is a numpy column vector containing the initial poles
        t is the time vector.
    Output:
        poles the poles in the system
        """

    # Initialize variables
    n = len(initPoles)  # The number of poles in the system
    timesteps = len(t)  # The number of timesteps in the input data

    xn = np.empty((n, timesteps), np.complex)
    yn = np.empty((n, timesteps), np.complex)

    # The vectors used in the convolution should have the same dimensions
    if x.shape != t.shape:
        x = np.reshape(x, t.shape)

    if y.shape != t.shape:
        y = np.reshape(y, t.shape)

    dt = t[1] - t[0]

    # Create waveforms from convolution
    for i in np.arange(n):
        xn[i] = np.convolve(np.exp(initPoles[i]*t), x)[0:timesteps]*dt
        yn[i] = np.convolve(np.exp(initPoles[i]*t), y)[0:timesteps]*dt

    # Prepare vectors for least squares
    A = np.concatenate((np.reshape(x, (timesteps, 1)),
                        np.ones((timesteps, 1))*2,
                        np.reshape(xn, (timesteps, n)),
                        np.reshape(-yn, (timesteps, n))), 1)
    y = np.reshape(y, (timesteps, 1))

    # Solve the system using least squares
    sol = np.linalg.lstsq(A, y, rcond=1e-4)

    # Find kn in the solution vector sol
    kn = sol[0][len(sol[0])-n:len(sol[0])]
    poles = findZeros(kn, initPoles)



def findZeros(kn, qn):
    """Function that finds the zeros in the fit function sigma"""

    [a, b] = sig.invres(kn, qn, np.array([1]))

    return np.roots(a)

File: MATLAB/findPoles.m
function [realPoles, complexPoles, err]...
    = findPoles(x, y, t, complexPoles, realPoles, tol)
% FINDPOLES find poles of a system using vector fitting
%
% INPUT:
%   x: Input signal
%   y: Output signal
%   t: time signal
%   complexPoles: The initial complex poles of the system. The code
%   calculates the conjugates
%   realPoles: The initial real poles of the system
%
% OUTPUT:
%   realPoles: The estimated real poles
%   complexPoles: The estimated complex poles

if nargin < 6
    tol = 1e-5;
end

if nargin < 5
    realPoles = [];
end

% Timesteps
ts = numel(t);

% Initialize some variables for convenience
kn = 0; % Residues for realPoles
knI = 0; % Real part of the residues for complexPoles
knII = 0; % Imaginary part of the residues for complexPoles

poleForward = [];

% Number of complexPoles
nC = numel(complexPoles);

% Number of real Poles
nR = numel(realPoles);

xnR = sparse(ts,nR);
ynR = sparse(ts,nR);

xnI = sparse(ts,nC);
ynI = sparse(ts,nC);
xnII = sparse(ts,nC);
ynII = sparse(ts,nC);


% Convolution between exponential of each pole and signals. Results are
% stored in separate columns for each pole
if ~isempty(realPoles)
    xnR = windowConv (x, realPoles, t);
    ynR = windowConv (y, realPoles, t);
end

if ~isempty(complexPoles)
    temp = windowConv (x, complexPoles, t);
    xnI = real(temp);
    xnII = imag(temp);

    temp = windowConv (y, complexPoles, t);
    ynI = real(temp);
    ynII = imag(temp);
end

A = [x, xnR, -ynR, 2*xnI, -2*xnII, -2*ynI, 2*ynII];

sol = (full(A)\y);

kn = sol(nR+2:1+2*nR)'; % Residues for real poles
knI = sol(end-2*nC+1:end-nC)'; % Real part of residues for complex poles
knII = sol(end-nC+1:end)'; % Imagiary part of residues for complex poles

% When normalizing the poles knI and knII cannot be 1x0 empty
if isempty(complexPoles)
    knI=[];
    knII=[];
    complexPoles = [];
end
if isempty(realPoles)
    kn = [];
    realPoles = [];
end
% Find the highest residue
abskn = [abs(kn)./abs(realPoles),abs(complex(knI,knII))./abs(complexPoles)];

err = norm(abskn);
    
% Check whether or not we already have the correct poles
if err < tol
    return
end
%Check if we are dealing with complex pairs
if nC > 0
%Create the Â matrix from Gustavsen paper
    temp = repmat(real(complexPoles.'),1,2)';
    AHat = diag(temp(:));
    %Indices of the superdiagonal
    v = 1:2:2*nC-1;
    w = 2:2:2*nC;
    superDiag = sub2ind([2*nC,2*nC], v, w); % Not really a superdiagonal
    AHat(superDiag) = imag(complexPoles);
    subDiag = sub2ind([2*nC,2*nC], w, v);
    AHat(subDiag) = -imag(complexPoles);
    bc = zeros(2*nC);
    bc(superDiag) = 2*knII;
    idx = sub2ind([2*nC,2*nC], 1:2:2*nC-1,1:2:2*nC-1);
    bc(idx) = 2*knI;
    temp = eig(AHat-bc)';
    
    % If there are real poles forward them
    poleForward = -abs(real(temp(abs(imag(temp))<tol)));
    temp = temp(abs(imag(temp))>tol);
    
    % Return only the negative pair of the complex conjugate pairs
    complexPoles = complex(-abs(real(temp(1:2:end))), -abs(imag(temp(1:2:end))));
    
    % If no complex poles were found return an empty variable instead of
    % and empty matrix
    if isempty(complexPoles)
        complexPoles =[];
    end
end
if nR > 0
    % Find the zeros of the fit function
     realPoles=eig(diag(realPoles)-ones(numel(kn),1)*kn)';
    % See if there are any complex poles
    if ~isreal(realPoles)
        % Order the complex poles into pairs
        temp = cplxpair(realPoles(abs(imag(realPoles))>tol));
        % Put the poles not already in complexPoles into complexPoles
        complexPoles = [complexPoles,...
            setdiff(complex(-abs(real(temp(1:2:end))),...
        -abs(imag(temp(1:2:end)))),complexPoles)];
        realPoles = -abs(realPoles(abs(imag(realPoles))<tol));
    else
        % Flip unstable poles
        realPoles(realPoles>0) = -realPoles(realPoles>0);
    end
end
temp = setdiff(poleForward,realPoles);
if ~isempty(temp)
    % In case the code for complex poles found any real poles.
    realPoles = [realPoles, temp];
end
